Fix synchronisation issue with setMediaOverridesForTesting state when a new WebContent process is used for testing

gavin-apple opened the pull request:
#### 419ca9b090186afea80666a5f59f586e89ef29e2
<pre>
Fix synchronisation issue with setMediaOverridesForTesting state when a new WebContent process is used for testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=253468">https://bugs.webkit.org/show_bug.cgi?id=253468</a>
rdar://106328795

Reviewed by Brady Eidson.

This changes resolves a synchronisation issue which arose when a new WebContent
process was spawned for a media-source test after a previous test changed
the setMediaOverridesForTesting from the default values.

The way setMediaOverridesForTesting works is that a singleton state is
maintained in both the WebProcess and the GPUP. When you request an
update to setMediaOverridesForTesting, the new state is checked against
the current local state stored within the process. If the two states are
different, a CoreIPC message is dispatched to the GPUProcess to tell
it to update the GPU side configuration.

WKTR will spawn new WebProcesses if the prior WebProcess is not suitable
for the next test, the prior WebProcess crashed, etc. When a new WebProcess
is spawned, the override&apos;s state is set back to it&apos;s default values.
However, we never reset the GPUP side state in this case. This would lead
to flaky failures.

During Internals construction, we now force reset WebProcess &amp; GPU side
state back to default values.

* Source/WebCore/platform/cocoa/SystemBattery.h:
* Source/WebCore/platform/cocoa/SystemBattery.mm:
(WebCore::SystemBatteryStatusTestingOverrides::setHasBattery):
(WebCore::SystemBatteryStatusTestingOverrides::setHasAC):
(WebCore::SystemBatteryStatusTestingOverrides::setConfigurationChangedCallback):
(WebCore::SystemBatteryStatusTestingOverrides::forceConfigurationUpdate):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::VP9TestingOverrides::setHardwareDecoderDisabled):
(WebCore::VP9TestingOverrides::setVP9DecoderDisabled):
(WebCore::VP9TestingOverrides::setVP9ScreenSizeAndScale):
(WebCore::VP9TestingOverrides::setConfigurationChangedCallback):
(WebCore::VP9TestingOverrides::forceConfigurationUpdate):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::Internals):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::updateMediaConfiguration):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setUseGPUProcessForMedia):

Canonical link: <a href="https://commits.webkit.org/261520@main">https://commits.webkit.org/261520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fdc9a86b642e9ee128401eeaebbc6508d219beed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111894 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3655 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120583 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12078 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3403 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117656 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16613 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99772 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104910 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98564 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/337 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45590 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13468 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/350 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/92704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13986 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9761 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19422 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52351 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8024 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15951 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

beidson reviewed (approved):
Type in commit message "When you requests..."

gavin-apple committed:
Fix synchronisation issue with setMediaOverridesForTesting state when a new WebContent process is used for testing
https://bugs.webkit.org/show_bug.cgi?id=253468
rdar://106328795

Reviewed by Brady Eidson.

This changes resolves a synchronisation issue which arose when a new WebContent
process was spawned for a media-source test after a previous test changed
the setMediaOverridesForTesting from the default values.

The way setMediaOverridesForTesting works is that a singleton state is
maintained in both the WebProcess and the GPUP. When you request an
update to setMediaOverridesForTesting, the new state is checked against
the current local state stored within the process. If the two states are
different, a CoreIPC message is dispatched to the GPUProcess to tell
it to update the GPU side configuration.

WKTR will spawn new WebProcesses if the prior WebProcess is not suitable
for the next test, the prior WebProcess crashed, etc. When a new WebProcess
is spawned, the override's state is set back to it's default values.
However, we never reset the GPUP side state in this case. This would lead
to flaky failures.

During Internals construction, we now force reset WebProcess & GPU side
state back to default values.

* Source/WebCore/platform/cocoa/SystemBattery.h:
* Source/WebCore/platform/cocoa/SystemBattery.mm:
(WebCore::SystemBatteryStatusTestingOverrides::setHasBattery):
(WebCore::SystemBatteryStatusTestingOverrides::setHasAC):
(WebCore::SystemBatteryStatusTestingOverrides::setConfigurationChangedCallback):
(WebCore::SystemBatteryStatusTestingOverrides::forceConfigurationUpdate):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::VP9TestingOverrides::setHardwareDecoderDisabled):
(WebCore::VP9TestingOverrides::setVP9DecoderDisabled):
(WebCore::VP9TestingOverrides::setVP9ScreenSizeAndScale):
(WebCore::VP9TestingOverrides::setConfigurationChangedCallback):
(WebCore::VP9TestingOverrides::forceConfigurationUpdate):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::Internals):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::updateMediaConfiguration):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setUseGPUProcessForMedia):

Canonical link: https://commits.webkit.org/261520@main
